### PR TITLE
fix(supervisor): reload env_file on restart and never orphan grandchildren (#29)

### LINKED ISSUE #29: restart should reload env_file and must not orphan grandchildren
## Summary

`prox restart <process>` can report success while continuing to run a grandchild from the previous process instance. It also reuses the environment captured when Prox first started instead of reloading configured `env_file` values.

Together these behaviors make a secret/configuration update appear to have been applied even though the old worker, with its old environment, can remain alive and retain its listening port.

Observed with Prox v0.1.2 on Linux.

## Observed behavior

A project uses a global environment file and a multi-process worker:

```yaml
env_file: backend/.env

processes:
  worker:
    cmd: cd backend && uv run --no-sync python -m app.restate.serve
```

1. Start Prox and the worker.
2. Change a value in `backend/.env`.
3. Run `prox restart worker`.
4. Prox reports that the process restarted, but the old `uv`/Python worker can survive with the old environment.
5. The replacement worker may fail with `EADDRINUSE` because the old Python process still owns the port.
6. `prox stop` shuts down the API/daemon but can leave the same worker behind. A subsequent `prox up` continues failing until the orphan is killed manually.

In the concrete occurrence, old `uv` and `python -m app.restate.serve` PIDs survived both restart and full stop. After those PIDs were killed and a genuinely fresh worker was started, `/proc/<pid>/environ` matched the current env file and the downstream request used the new configuration successfully.

No credential values are included in this report.

## Likely root cause

There appear to be two independent issues which compound one another:

1. **Environment files are cached for the supervisor lifetime.** `LoadProcessEnv` is called only by `Supervisor.createManagedProcess()` when the supervisor starts. `RestartProcess()` calls `ManagedProcess.Restart()`, which stops and starts the same object with its existing `env` map. It never reloads either the global or process-specific `env_file`.

2. **Shutdown completion follows the command/group leader, not the entire process group.** Prox starts `sh -c ...` as a process-group leader and sends signals to its group. However, `monitor()` waits for the `exec.Cmd` leader only. If that leader exits after SIGTERM while a grandchild remains, output draining times out after five seconds and the managed process is marked `stopped`; this closes `done`, so `Stop()` returns before its shutdown timeout and never sends SIGKILL. Subsequent signaling derives the PGID from the exited leader, so the surviving group may no longer be reachable through the stored process object.

Relevant code:

- `internal/supervisor/supervisor.go`: `createManagedProcess`, `RestartProcess`
- `internal/supervisor/process.go`: `Stop`, `Restart`, `monitor`
- `internal/supervisor/runner.go`: `ExecRunner.Start`, `execProcess.Signal`

## Expected behavior

- Starting or restarting a process reloads configured global and process-specific `env_file` values.
- Restart does not start the replacement until the previous process group is gone.
- Stop/restart treats the complete process group as the lifecycle boundary.
- If any descendant remains after the graceful timeout, Prox sends SIGKILL to the original process group and verifies it is gone.
- A full `prox stop` does not leave managed descendants running.
- If cleanup cannot be completed, stop/restart returns an error rather than reporting success.

## Suggested tests / acceptance criteria

- Start a process that reports a value from an env file, change the file, run `prox restart`, and assert the new process reports the new value.
- Start a wrapper that spawns a grandchild which ignores SIGTERM while its parent exits. Assert `prox stop <name>` kills the grandchild after the timeout.
- Run the same stubborn-grandchild case through `prox restart`; assert the old group is absent before the replacement starts and the listening port can be rebound.
- Run it through full `prox stop`; assert no process in the original group remains.
- Preserve the original PGID at process start instead of resolving it only from a potentially exited leader.

A possible implementation direction is to load environment data on every start attempt (or recreate the managed process on restart), store the PGID when the command launches, and gate stop/restart completion on process-group disappearance rather than only `cmd.Wait()` for the shell leader.

### PR DESCRIPTION
**Pull request by charliek**
Fixes #29.

`prox restart` could report success while (a) running the replacement with a **stale environment** and (b) leaving the previous process's **grandchildren alive**, so an updated secret/config appeared applied while the old worker kept its listening port (`EADDRINUSE`). `prox stop` could leave the same orphan and still report success. Two independent root causes are fixed here, each as its own commit.

### 1. `env_file` is reloaded on every start (`dfc93ce`)
`config.LoadProcessEnv` was called once at supervisor start and the result cached for the supervisor's lifetime. Env loading now happens in a `loadEnv` closure invoked at the top of `ManagedProcess.Start` (before any lifecycle state is created), so `start`-after-`stop` and the replacement half of `restart` both re-read the global + process `env_file` and re-merge inline `env`. A failing reload fails loudly (process marked crashed, `Start` returns a wrapped `ErrEnvReloadFailed`) instead of launching with a stale environment.

### 2. Restart starts the replacement on the supervisor context (`a36de9a`)
`RestartProcess` threaded the request-scoped context into the replacement's `Start`, so once the HTTP handler returned, the fresh process's health checker and lifecycle context were torn down. `Restart` now takes distinct stop/start contexts and starts the replacement on the supervisor's long-lived `s.ctx` (matching `StartProcess`).

### 3. PGID captured at launch + group-liveness probe (`4faf583`)
The process group id is recorded at launch (the leader PID, since we `Setpgid`) instead of being re-derived via `Getpgid` at signal time (which fails once the leader exits, dropping the surviving group). Adds `Process.GroupAlive() (bool, error)` — a signal-0 probe — and hard-guards group signaling to `pgid > 0` so it can never signal prox's own group.

### 4. Stop/restart gate on the whole process group and kill orphans (`332ad8b`)
- Per-run `processInstance` with a **generation guard**, so a stale `monitor` from a previous run can never clobber a newer run; the live instance (and its pgid) is retained so a surviving group stays reapable.
- `Stop` now: SIGTERM the group → poll `GroupAlive()` until a **time-based graceful deadline** (never keyed off the leader's exit, so a slow-but-graceful grandchild isn't killed early) → **SIGKILL the group** at the deadline → finalization gate (wait for the run's monitor to finish and reap the leader) → authoritative re-probe. Group gone ⇒ `stopped`; group survived ⇒ `crashed` + `ErrProcessGroupNotReaped`.
- **Failure contract:** `prox stop <name>` / `prox restart <name>` now return a non-zero exit / typed API error (`PROCESS_GROUP_NOT_REAPED`) when a group can't be reaped, instead of always reporting success (matches the systemd/docker convention). Full `prox stop` best-effort-kills each group and logs `could not reap process group for <name>`.

## Verification

- **`make lint` (0 issues), `make build`, `make test`, `go test -race ./internal/supervisor/...` green.**
- **Unit tests** (deterministic, via a fake `ProcessRunner`/`Process`, run under `-race`): graceful death → success + *no* SIGKILL; stubborn → SIGKILL escalation → success; unreapable → `ErrProcessGroupNotReaped` + retry; restart-clobber loop; `Start`-during-`Stop` rejection; deadline edge cases; PGID capture / `GroupAlive`; env reload + typed failure; restart health-checker survival.
- **Integration tests** (real binary, real processes/ports): a stubborn grandchild (Python `SIG_IGN` on SIGTERM holding a TCP port, under a leader shell that exits on SIGTERM) is killed by `stop`, its port rebinds after `restart`, and full `stop` leaves no orphan; env reload is exercised end-to-end through an edited env file. Each stubborn test kills recorded grandchild PIDs in `t.Cleanup` (no CI orphan leak). The pre-existing graceful-shutdown output-capture test still passes (no premature SIGKILL).
- **Manual, real CLI:** `prox restart` picked up an edited `env_file` value (exit 0); `prox stop worker` reaped a stubborn orphan after the ~8s graceful window (log: `sending SIGKILL to worker (graceful shutdown timed out)`), returned exit 0, and freed the port.

## Impact / notes

- **Behavior change:** `stop`/`restart` can now take up to the graceful window (~8s) before SIGKILLing a stuck group, and `prox stop <name>` / `prox restart <name>` can now return a **non-zero exit** when a group genuinely can't be reaped. Scripts that assumed `stop` always succeeds should be aware.
- **No new dependencies. No secrets touched** — env values are never logged (only the failing env-file path is surfaced in an error, which is the user's own config path). Unix-only, consistent with the existing supervisor.
- **Accepted/known limitations (documented):** pgid-reuse window and `setsid`/detached-group escape (inherent to pgid-based signaling; a cgroup-scoped kill is future work); PID-1-without-reaping zombie lingering; a concurrent same-process `Stop` may report success just before the primary's verdict (only inconsistent given a genuinely unreapable group); no-arg `prox stop` reports async and best-effort. Reloading a changed `cmd`/`prox.yaml` on restart is intentionally **out of scope** (env values only) — a good follow-up.

## Process

Plan authored, pressure-tested by an AI panel (Codex + Kimi + CodeRabbit), then implemented as four gated commits (lint/test/build + `-race` per commit), each simplified and correctness-reviewed by an adversarial Codex pass. Two Codex findings on the concurrency were dispositioned: the finalization-gate/monitor state-override race was **fixed**; the concurrent-same-process-`Stop` edge was **documented** as future work.

<details>
<summary>Plan excerpt — pinned design decisions (D1–D4)</summary>

**D1 — Reload `env_file` inside every `Start`:** a `loadEnv` closure captures the env-file paths + inline map + configDir by value and re-reads on each start; failure ⇒ crashed + `ErrEnvReloadFailed` with no dangling instance.

**D2 — Capture PGID at launch; group-aware probe:** store `cmd.Process.Pid` as the pgid; `GroupAlive() (bool,error)` maps ESRCH/`os.ErrProcessDone` ⇒ gone, nil/EPERM/other ⇒ conservatively alive; hard `pgid>0` guards in `Signal`/`GroupAlive`.

**D3 — Per-run instance model + group-gated stop:** `processInstance{proc,done,cancel,outputWg}`; `current` retained; generation-guarded `monitor` (only `running→crashed`, never overrides a terminal state); `Stop` = SIGTERM → time-based graceful poll → SIGKILL at `gracefulDeadline` → finalization gate on `inst.done` → verdict; `computeDeadlines` reserves `KillGrace` (2s) inside `ShutdownTimeout` and handles no-deadline / near-expired ctx; `Start` rejects `stopping` and refuses to launch over a surviving group.

**D4 — Failure contract:** `ErrProcessGroupNotReaped`/`PROCESS_GROUP_NOT_REAPED` + `writeError` case ⇒ non-zero CLI exit; full stop best-effort + logs the un-reaped group.

Acceptance criteria A1–A11 map 1:1 to these workstreams and are covered by the unit + integration + manual verification above.
</details>

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved process stopping/restarting to reliably handle stubborn processes and prevent stale monitors from impacting newer runs.
  * Added stronger cleanup checks to ensure process groups are fully reaped (with dedicated error codes when reaping fails).
  * Reloads environment files on every start and rejects start attempts while a stop is still in progress.
* **Tests**
  * Expanded unit tests for stop/restart deadlines, error mapping, and start-during-stop protection.
  * Added integration tests covering stubborn grandchild shutdown, port rebinds on restart, env reload, and no orphaned processes.

### CODE REVIEW

File: internal/supervisor/process.go [hunk 1]
```diff
@@ -155,186 +219,304 @@ func (p *ManagedProcess) Start(ctx context.Context) error {
...
 		// Output readers finished normally
 	case <-time.After(outputDrainTimeout):
-		p.logManager.Write(domain.LogEntry{
-			Timestamp: time.Now(),
-			Process:   p.config.Name,
-			Stream:    domain.StreamStderr,
-			Line:      "output capture timed out (some logs may be missing)",
-		})
+		p.logf(domain.StreamStderr, "output capture timed out (some logs may be missing)")
 	}
```
[the remark this answers is not in the document]
Comment (charliek) [reply internal/supervisor/process.go:462]: Addressed in 3948a46. The drain-timeout log is now gated by the generation guard — rather than the suggested `RLock` in the select branch, it defers the log into the existing `p.mu.Lock` block via a `drainTimedOut` flag, so a stale run (`p.current != inst`) emits neither the drain-timeout notice nor the exit-code log, without a second lock acquisition.

File: internal/supervisor/supervisor.go [hunk 2]
```diff
@@ -366,17 +376,22 @@ func (s *Supervisor) StopProcess(ctx context.Context, name string) error {
...
+	// Create timeout context to bound the stop half of the restart.
 	restartCtx, cancel := context.WithTimeout(ctx, s.supConfig.ShutdownTimeout)
 	defer cancel()
 
-	err := mp.Restart(restartCtx)
+	// Stop uses restartCtx (bounded by the request/shutdown timeout); Start
+	// uses supCtx so the replacement's process lifecycle and health checker
+	// survive after this request's context is cancelled/expires (mirrors
+	// StartProcess).
+	err := mp.Restart(restartCtx, supCtx)
```
[the remark this answers is not in the document]
Comment (charliek) [reply internal/supervisor/supervisor.go:409]: Addressed in 3948a46. Both `StartProcess` and `RestartProcess` now guard `supCtx == nil` (returning `ErrShutdownInProgress`) before it can reach `context.WithCancel`, matching the nil guard in `Wait()`.

File: test/integration/helpers_test.go [hunk 3]
```diff
@@ -138,6 +138,49 @@ func stopProx(t *testing.T, addr string) error {
 	return nil
 }
 
+// restartProcess sends a POST /api/v1/processes/{name}/restart request and
+// returns the parsed error response (Code/Error) when the request did not
+// return 200, or a zero-value response and nil error on success.
+func restartProcess(t *testing.T, addr, name string) (int, ErrorResponse) {
+	t.Helper()
+	return postProcessAction(t, addr, name, "restart")
+}
```
[the remark this answers is not in the document]
Comment (charliek) [reply test/integration/helpers_test.go:147]: Addressed in 3948a46. `restartProcess`'s doc now reads "returns the HTTP status code and the parsed error response (empty Code/Error on success)", matching its `(int, ErrorResponse)` signature.

**Comment by charliek** [comment]
Follow-ups identified during this work, now tracked as issues:

**Bugs**
- #31 — healthcheck `interval`/`timeout`/`retries`/`start_period` silently ignored (Medium/High severity, small effort)
- #32 — concurrent same-process stop can report false success (Low, small–medium) — the limitation documented inline in `Stop` here

**Enhancements**
- #33 — `restart` should hot-reload a changed `cmd`/config (High value, medium–large) — the env-only scoping in this PR
- #34 — cgroup-scoped termination on Linux (Medium, large) — closes the pgid-reuse / `setsid` / PID-1-zombie limitations documented here
- #35 — configurable graceful/kill timeout (Medium, small–medium)
- #36 — full `prox stop` should report per-process reap failures (Low/Medium, medium) — extends this PR's per-process failure contract
